Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,446 @@
+import Batteries.Data.Array
+import Batteries.Data.List.Basic
+import Batteries.Data.List.Lemmas
+
+namespace Vector
+
+open Array Lean
+
+/--
+  `Vector α n` is an `Array α` whose size is statically fixed to `n`
+-/
+structure Vector (α : Type u) (n : Nat) where
+  /--Internally, a vector is stored as an array for fast access-/
+  toArray : Array α
+  /--`size_eq` fixes the size of `toArray` statically-/
+  size_eq : toArray.size = n
+deriving Repr, BEq, DecidableEq
+
+
+/--Syntax for `Vector α n`-/
+syntax "#v[" withoutPosition(sepBy(term, ", ")) "]" : term
+
+macro_rules
+  | `(#v[ $elems,* ]) => `(Vector.mk (n := $(quote elems.getElems.size)) #[$elems,*] rfl)
+
+#check #v[1,2,3,3,4,4]
+
+/--
+  `Vector.size` gives the size of a vector.
+-/
+def size (_ : Vector α n) : Nat := n
+
+
+/--
+  `Vector.nil` produces an empty vector
+-/
+@[match_pattern]
+def nil : Vector α 0 := ⟨Array.empty, by rfl⟩
+
+/--
+  `Vector.cons` is the list cons pattern for vectors
+-/
+@[match_pattern]
+def cons (x : α) (xs : Vector α n) : Vector α (Nat.succ n) :=
+  match xs with
+  | ⟨{data := xs'}, h⟩ => ⟨{data := x :: xs'}, congrArg Nat.succ h⟩
```

Comment:
It seems to me that these can't work as match patterns.

Reply:
This one is a legacy of Mathlib.Vector

Reply:
I should probably remove them entirely.

Reply:
I agree: `nil` is just `#v[]` and `cons` is just `#v[x] ++ xs`.

Reply:
Done. Moved the function below append. Added an HAppend instance.